File: _Demo_and_Tests/Objective-C/AlgorithmsTests/QuickSortTests.m
//
//  QuickSortTests.m
//  Algorithms
//

#import <XCTest/XCTest.h>
#import "QuickSort.h"
#import "TestsHelper.h"

@interface QuickSortTests : XCTestCase

@end

@implementation QuickSortTests

- (void)setUp {
    [super setUp];

    [TestsHelper reset];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark -

- (void)testAscOdd {
    NSArray *resultArray = [QuickSort sortedArray:TestsHelper.array order:QuickSortOrderAsc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayAsc);
}

- (void)testAscEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayAsc addObject:@100];

    NSArray *resultArray = [QuickSort sortedArray:TestsHelper.array order:QuickSortOrderAsc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayAsc);
}

- (void)testDescOdd {
    NSArray *resultArray = [QuickSort sortedArray:TestsHelper.array order:QuickSortOrderDesc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayDesc);
}

- (void)testDescEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayDesc insertObject:@100 atIndex:0];

    NSArray *resultArray = [QuickSort sortedArray:TestsHelper.array order:QuickSortOrderDesc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayDesc);
}

- (void)testNilArray {
    NSArray *resultArray = [QuickSort sortedArray:nil order:QuickSortOrderAsc];

    XCTAssert(resultArray == nil);
}

- (void)testEmptyArray {
    NSArray *resultArray = [QuickSort sortedArray:@[] order:QuickSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[]);
}

- (void)testArrayWithOneNumber {
    NSArray *resultArray = [QuickSort sortedArray:@[@0] order:QuickSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[@0]);
}

@end

File: _Demo_and_Tests/Swift/AlgorithmsTests/QuickSortTests.swift
//
//  QuickSortTests.swift
//  Algorithms
//

import XCTest

class QuickSortTests: XCTestCase {

    override func setUp() {
        super.setUp()

        TestsHelper.reset()
    }

    override func tearDown() {
        super.tearDown()
    }

    // MARK: -

    func testAscOdd() {
        let result = QuickSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(result ?= TestsHelper.sortedArrayAsc)
    }

    func testAscEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayAsc.append(100)

        let result = QuickSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(result ?= TestsHelper.sortedArrayAsc)
    }

    func testDescOdd() {
        let result = QuickSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(result ?= TestsHelper.sortedArrayDesc)
    }

    func testDescEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayDesc.insert(100, at: 0)

        let result = QuickSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(result ?= TestsHelper.sortedArrayDesc)
    }

    func testNilArray() {
        let nilArray: [Int]? = nil
        let result = QuickSort.sort(nilArray, order: .ASC)

        XCTAssert(result == nil);
    }

    func testEmptyArray() {
        let result = QuickSort.sort([] as [Int], order: .ASC)

        XCTAssert(result ?= []);
    }

    func testArrayWithOneNumber() {
        let result = QuickSort.sort([0], order: .ASC)
        
        XCTAssert(result ?= [0]);
    }

}
